### PR TITLE
Added code to reverse transforms for more flexibility

### PR DESCRIPTION
**Pull request by ThibeauWouters**
New attempt at improving the flexibility and use of "pre-made" transforms such as masses transforms etc. 

**Source code changes**

I have added two utility functions: 
- One function to create a `BijectiveTransform` given a `name_mapping` and the forward and backward transformation functions which take arrays as input. That way, we can easily create these "pre-made" transforms by specifying these in a one-liner. At the same time, it allows users to create similar transforms themselves very easily, which therefore also makes the `name_mapping` no longer hardcoded for the masses since users can change it if they desire, as @thomasckng mentioned some time ago.
- Another function reverses a `BijectiveTransform` object, as @kazewong suggested in my previous PR attempt. 

These functions are then used to simplify the "pre-made" transforms in `single_event/transforms.py` to be shorter and clearer.

**Test changes**

- Made sure the GW150914_D.py file in `test/integration` matches this new implementation.
- Added a new toy example test file to play around with the transforms in a simplified setting.

### CODE REVIEW

**Comment by thomasckng** [comment]
@ThibeauWouters I think the test script in `test/` are meant to be light-weighted, as they will run whenever there is a new commit. They meant to check if everything is working instead of generating meaningful result. A example with optimized parameters are meant to be put in `example/`. I am canceling the test runs now.

**Comment by ThibeauWouters** [comment]
@thomasckng Thanks! I had local copies of the test file to try more agressive hyperparameters but indeed I seemed to have messed them up with the ones in the test directory. Thanks for canceling the runs.

**Comment by xuyuon** [comment]
@ThibeauWouters I am planning to merge my PR with your PR. There are some conflicts in the test directory. I'll be resolving these conflicts by overwriting your files in the test directory with my files. Please let me know if there is still some changes in your test directory that you wish to preserve.

**Comment by xuyuon** [comment]
@ThibeauWouters I have created a PR to merge my commits into your PR. Could you please take a look at it to see if there is any issue?

File: test/integration/test_GW150914_D.py [hunk 1]
```diff
@@ -97,34 +103,39 @@
...
-    n_chains=4,
+    n_loop_production=4,
+    n_local_steps=10,
+    n_global_steps=1000,
+    n_chains=500,
     n_epochs=n_epochs,
     learning_rate=learning_rate,
-    n_max_examples=30,
-    n_flow_samples=100,
+    n_max_examples=30000,
```
Comment (kazewong) [inline test/integration/test_GW150914_D.py:127]: Test are supposed to be light weight. Reminder to revert these numbers

File: src/jimgw/transforms.py [hunk 2]
```diff
@@ -445,3 +445,54 @@ def __init__(
             )
             for i in range(len(name_mapping[1]))
         }
+
+
+def create_bijective_transform(
```
Comment (kazewong) [inline src/jimgw/transforms.py:450]: I think this is not a good practice in general. Here `transform_func_array` and `inverse_transform_func_array` can be potentially harmful functions, and in the long run, this flexibility will create trouble in reproducibility. Any bijective transform should be explicitly declared with the class interface.

**Comment by ThibeauWouters** [comment]
@kazewong I adapted the code to adapt to your previous comment. But now it feels a bit stupid in the sense that the "predefined" transforms are classes, that have to be instantiated in order to get the actual transform to be used in PE, and the reverse transforms that are created from the reverse function are objects/instances already. The former require users to give `name_mapping`, which as I mentioned earlier on is a bit redundant since we would like the name mapping to be fixed for these classes, whereas the former doesn't. So the predefined transform and its reverse are not treated equally in the current implementation.

File: src/jimgw/single_event/transforms.py [hunk 3]
```diff
@@ -20,45 +24,70 @@
 
 
 @jaxtyped(typechecker=typechecker)
-class ComponentMassesToChirpMassMassRatioTransform(BijectiveTransform):
+class SpinToCartesianSpinTransform(NtoNTransform):
```
Comment (kazewong) [inline src/jimgw/single_event/transforms.py:27]: I see a lot of new transforms in the code below. Are they relevant to this PR? If not, can we separate them into a different PR for traceability?

**Comment by kazewong** [comment]
> @kazewong I adapted the code to adapt to your previous comment. But now it feels a bit off in the sense that the "predefined" transforms are classes, that have to be instantiated in order to get the actual transform to be used in PE, and the reverse transforms that are created from the reverse function are objects/instances already. The former require users to give `name_mapping`, which as I mentioned earlier on is a bit redundant since we would like the name mapping to be fixed for these classes, whereas the former doesn't. So the predefined transform and its reverse are not treated equally in the current implementation.

There are a couple of things you mentioned here:
1. Re: predefined transform and its reverse are not treated equally: I don't think they should. Inverting a transform target a bijective transform, and predefined transforms are a subset of bijective transforms. For example, `reverse_bijective_transform(ScaleTransform((['x'], ['y']), 1.0))` should return a valid transform. As long as this is the interface the user will use, I think this is legit.
2. If we later establish predefined transform can be called with `PredefinedTransform()`, the interface makes sense to me still, which will be replacing `PredefinedTransform()` with `reverse_bijective_transform(PredefinedTransform())` at the same spot.

I think this is more ergonomic for the user, instead of needing to use a function to call the predefined transform.

Going forward, I do think predefined transforms that enforce the naming strictly should mostly be constructed using the interface `PredefinedTransform()`. Only transform that does not enforce naming such as `ScaleTransform` should allow inputs other than the transform parameters.

File: src/jimgw/single_event/transforms.py (hunk 3 above)
Comment (ThibeauWouters) [reply src/jimgw/single_event/transforms.py:27]: These are no new transforms.I just might have shuffled around their positions a bit (by accident) which is why they show up in the diffs.

**Comment by ThibeauWouters** [comment]
@thomasckng @kazewong I have updated the PR a bit after the discussion yesterday. Let me know if this looks OK to you or if further changes are required.

File: src/jimgw/single_event/transforms.py (hunk 3 above)
Comment (kazewong) [reply src/jimgw/single_event/transforms.py:27]: Okay. We can proceed then

File: test/integration/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+outdir/
+figures/
```
Comment (thomasckng) [inline test/integration/.gitignore:2]: Figures are still uploaded to the repo now.

**Comment by ThibeauWouters** [comment]
@thomasckng The files have been deleted

**Review by thomasckng — APPROVED** [review]

**Review by kazewong — COMMENTED** [review]
Please resolve the comment related to further consolidation. But after that it should be good for merging

**Review by xuyuon — APPROVED** [review]

**Comment by ThibeauWouters** [comment]
@kazewong I have updated the source code as requested -- is this what you were referring to?

**Review by kazewong — APPROVED** [review]
